src/core/service.c

```diff
@@ -1093,6 +1094,20 @@ static void service_dump(Unit *u, FILE *f, const char *prefix) {
                         fprintf(f, "%sOpen File: %s\n", prefix, ofs);
                 }
 
+        LIST_FOREACH(fds, extra_fd, s->extra_fds) {
```

Comment:
Hmm, the extra fds are never released even after Service stops? Please refer to `service_release_stdio_fd` for proper cleanup upon exit.

Reply:
This needs to be serialized too BTW. Check out how it is done for fd store fds.

Reply:
> Hmm, the extra fds are never released even after Service stops? Please refer to service_release_stdio_fd for proper cleanup upon exit.

They are freed in line 513 above in service_done() via extra_file_descriptor_free_many() - which will release memory and close FDs

Looking at service_release_stdio_fd, I also forgot to add this in service_release_resources(), I will add it there, thx for catching!

> This needs to be serialized too BTW. Check out how it is done for fd store fds.

To be clear, you want me to do more detailed serialization like fdstore (e.g. path/device/etc)? https://github.com/systemd/systemd/blob/615279f369aa4116e21704386090570dc42f4a55/src/core/service.c#L929

Reply:
Fixed both! I also refactored service_dump_fdstore() to reuse logic with service_dump_extra_fds()